src/mds/MDLog.cc

```diff
@@ -603,11 +614,33 @@ void MDLog::try_to_commit_open_file_table(uint64_t last_seq)
   }
 }
 
-void MDLog::trim(int m)
+void MDLog::log_trim_upkeep(void) {
+  dout(10) << dendl;
+
+  submit_mutex.lock();
```

Comment:
I don't think introducing the submit mutex here just to wait on the submit condition is justified. Submit mutex is already overused in MDLog, which I addressed in the refactoring PR, but here it's also not required and hence is misleading IMO.

MDS lock is sufficient for the trim method, and in order to wake up the thread when shutting down, a signal can be sent to the thread that would be sleeping in a `sigtimedwait` for that purpose - instead of the condition timed wait.

`upkeep_log_trim_shutdown` is already atomic so no additional locks are needed to test it.

Reply:
(sorry for jumping into this rather late)

> I don't think introducing the submit mutex here just to wait on the new condition is justified. Submit mutex is already overused in MDLog, which I addressed in the refactoring PR, but here it's also not required and hence is misleading IMO.

MDLog::trim() acquires submit mutex, so that is maintained by acquiring submit_mutex here and calling into MDLog::trim() (which does not acquire submit_mutex anymore with this change).

> 
> MDS lock is sufficient for the trim method, and in order to wake up the thread when shutting down, a signal can be sent to the thread that would be sleeping in a `sigtimedwait` for that purpose - instead of the condition timed wait.

Since submit_mutex is already taken so its used to wait on a condvar.

> 
> `upkeep_log_trim_shutdown` is already atomic so no additional locks are needed to test it.

Reply:
After looking at the code again, I think @leonid-s-usov is correct here. The `submit_mutex` is supposed to protect the `MDLog::pending_events` member (it should have a comment explaining that). The rest of `MDLog` is protected by the `mds_lock`.

I do not think `sigtimedwait` is necessary here. The `log_trim_upkeep` thread can have a condition variable associated with the `mds_lock` which can be signaled normally.

Note: the `MDLog::_submit_thread` does twiddle bits in the `LogSegment` associated with the `LogEvent` (evil!) but I think it's harmless that that access is not protected by the `mds_lock` (nothing else modifies those bits). As long as the `LogSegment` is not expired (and it cannot be with pending events), then the access is safe. So, indeed, only `MDLog::pending_events` needs protected by the `submit_mutex`. I think @leonid-s-usov 's refactor will address that fully. Let's work on getting that merged quickly once this is.

Reply:
this is the place where we suggested that you didn't need to lock the submit mutex, and this comment is still valid. Just use the MDS lock to guard the condition variable, and leave the submit mutex operation unchanged by this PR (i.e. it will continue to be taken by the `trim` function below)

The other PR about mdlog refactoring will address the usage of the submit mutex in the `trim` method